cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/security/jwt/JwtLoginService.java

```diff
@@ -0,0 +1,197 @@
+ * this class, so opening and closing connections should be done by implementing the {@link AbstractLifeCycle} interface's
+ * {@link #doStart()} and {@link #doStop()} methods respectively. For a simple example see
+ * {@link UserStoreAuthorizationService}.</p>
+ */
+public class JwtLoginService extends AbstractLifeCycle implements LoginService {
+
+  private final AuthorizationService _authorizationService;
+  private IdentityService _identityService = new DefaultIdentityService();
+  private final RSAPublicKey _publicKey;
+  private final List<String> _audiences;
+
+  public JwtLoginService(AuthorizationService authorizationService, String publicKeyLocation, List<String> audiences)
+      throws IOException, CertificateException {
+    this(authorizationService, readPublicKey(publicKeyLocation), audiences);
+  }
+
+  public JwtLoginService(AuthorizationService authorizationService, RSAPublicKey publicKey, List<String> audiences) {
+    _authorizationService = authorizationService;
+    _publicKey = publicKey;
+    _audiences = audiences;
+  }
+
+  @Override
+  protected void doStart() throws Exception {
+    super.doStart();
+    // The authorization service might want to start a connection or access a file
+    if (_authorizationService instanceof LifeCycle) {
+      ((LifeCycle) _authorizationService).start();
+    }
+  }
+
+  @Override
+  protected void doStop() throws Exception {
+    if (_authorizationService instanceof LifeCycle) {
+      ((LifeCycle) _authorizationService).stop();
+    }
+    super.doStop();
+  }
+
+  @Override
+  public String getName() {
+    return null;
+  }
+
+  @Override
+  public UserIdentity login(String username, Object credentials, ServletRequest request) {
+    if (!(credentials instanceof SignedJWT)) {
+      return null;
+    }
+    if (!(request instanceof HttpServletRequest)) {
+      return null;
+    }
+
+    SignedJWT jwtToken = (SignedJWT) credentials;
+    boolean valid;
+    try {
+      valid = validateToken(jwtToken, username);
+    } catch (ParseException e) {
+      JWT_LOGGER.warn(String.format("%s: Couldn't parse a JWT token", username), e);
+      return null;
+    }
+    if (valid) {
+      String serializedToken = (String) request.getAttribute(JwtAuthenticator.JWT_TOKEN_REQUEST_ATTRIBUTE);
+      UserIdentity rolesDelegate = _authorizationService.getUserIdentity((HttpServletRequest) request, username);
+      return getUserIdentity(jwtToken, serializedToken, username, rolesDelegate);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public boolean validate(UserIdentity user) {
+    return false;
+  }
+
+  @Override
+  public IdentityService getIdentityService() {
+    return _identityService;
+  }
+
+  @Override
+  public void setIdentityService(IdentityService service) {
+    _identityService = service;
+  }
+
+  @Override
+  public void logout(UserIdentity user) {
+
```

Comment:
This logout method I think is mostly used to implement the user initiated logout (when they click on the button) and is common with session based authentications. In this case this is a bit different because the token is issued by a third party so we need to call that to invalidate the token. Unfortunately Knox doesn't really provide a solid way for this. There is a proposal for this [KIP-10](https://cwiki.apache.org/confluence/display/KNOX/KIP-10+KnoxSSO+Logout+Flow) and some traces of this in their code but it doesn't seem to be a public and documented API so I didn't want to make a generic implementation based on that.

Reply:
One more addition: reading up on it, JWTs can't be easily expired as they're stateless. On the provider side it's possible to set an expiration time for the cookie.
A good article here: https://medium.com/devgorilla/how-to-log-out-when-using-jwt-a8c7823e8a6

Reply:
I see, cool :)